inc/functions/files.php

```diff
@@ -1031,72 +1031,82 @@ function rocket_clean_cache_dir() {
  * Remove a single file or a folder recursively
  *
  * @since 1.0
+ * @since 3.5.3 Bails if given dir should be preserved; replaces glob; optimizes.
  *
- * @param string $dir File/Directory to delete.
+ * @param string $dir              File/Directory to delete.
  * @param array  $dirs_to_preserve (default: array()) Dirs that should not be deleted.
- * @return void
  */
-function rocket_rrmdir( $dir, $dirs_to_preserve = [] ) {
-	$dir = untrailingslashit( $dir );
+function rocket_rrmdir( $dir, array $dirs_to_preserve = [] ) {
+	$dir        = untrailingslashit( $dir );
+	$filesystem = rocket_direct_filesystem();
+
+	// Bail out if the given directory is in the list of directories to preserve.
+	if ( ! empty( $dirs_to_preserve ) && $filesystem->is_dir( $dir ) && in_array( $dir, $dirs_to_preserve, true ) ) {
+		return;
+	}
 
 	/**
 	 * Fires before a file/directory cache is deleted
 	 *
 	 * @since 1.1.0
 	 *
-	 * @param string $dir File/Directory to delete.
-	 * @param array $dirs_to_preserve Directories that should not be deleted.
-	*/
+	 * @param string $dir              File/Directory to delete.
+	 * @param array  $dirs_to_preserve Directories that should not be deleted.
+	 */
 	do_action( 'before_rocket_rrmdir', $dir, $dirs_to_preserve ); // phpcs:ignore WordPress.NamingConventions.PrefixAllGlobals
 
 	// Remove the hidden empty file for mobile detection on NGINX with the Rocket NGINX configuration.
 	$nginx_mobile_detect_file = $dir . '/.mobile-active';
 
-	if ( rocket_direct_filesystem()->is_dir( $dir ) && rocket_direct_filesystem()->exists( $nginx_mobile_detect_file ) ) {
-		rocket_direct_filesystem()->delete( $nginx_mobile_detect_file );
+	if ( $filesystem->is_dir( $dir ) && $filesystem->exists( $nginx_mobile_detect_file ) ) {
+		$filesystem->delete( $nginx_mobile_detect_file );
 	}
 
 	// Remove the hidden empty file for webp.
 	$nowebp_detect_file = $dir . '/.no-webp';
 
-	if ( rocket_direct_filesystem()->is_dir( $dir ) && rocket_direct_filesystem()->exists( $nowebp_detect_file ) ) {
-		rocket_direct_filesystem()->delete( $nowebp_detect_file );
+	if ( $filesystem->is_dir( $dir ) && $filesystem->exists( $nowebp_detect_file ) ) {
+		$filesystem->delete( $nowebp_detect_file );
 	}
 
-	if ( ! rocket_direct_filesystem()->is_dir( $dir ) ) {
-		rocket_direct_filesystem()->delete( $dir );
+	if ( ! $filesystem->is_dir( $dir ) ) {
+		$filesystem->delete( $dir );
+
 		return;
-	};
+	}
 
-	$dirs = glob( $dir . '/*', GLOB_NOSORT );
-	if ( $dirs ) {
+	// Get the directory entries.
+	try {
+		$entries = new FilesystemIterator( $dir, FilesystemIterator::SKIP_DOTS );
+	} catch ( Exception $e ) {
+		$entries = [];
+	}
+
+	foreach ( $entries as $entry ) {
```

Comment:
Shouldn't we check if `$entries` is empty, before the `foreach`? And if empty, delete the directory then bailout?

Reply:
We are. If the `$entries` are empty, the `foreach` doesn't run and skips over all the code within the `foreach`. What happens next?

- It jumps down to line 1100 and deletes the directory.
- It then fires the `after_rocket_rrmdir` action event.
- It then exits the function. Done.

@Tabrisrp Do you think it shouldn't fire the action event?

Reply:
I was under the impression that the `foreach` would try to run even with an empty array and create an error, but it's very likely my brain doesn't work this morning

Reply:
Nope. If the array is empty, a `foreach` does not run.

Reply:
A new day a new thing to learn

Reply:
For PHP array functions, such as `foreach`, `array_walk`, etc., first check if there's anything to iterate over. If no, i.e. the array is empty, then these functions bail out. Pretty cool!